claasp/cipher_modules/continuous_diffusion_analysis.py

```diff
@@ -0,0 +1,613 @@
+                outer_log = (inner_log + Decimal('1')).ln() / Decimal('2').ln()
+                return outer_log
+            return 0
+
+        x_inputs = [
+            x if input_element == [] else input_element for input_element in lst_input
+        ]
+        sum_by_round = []
+
+        output_evaluated_continuous_diffusion_analysis_x = \
+            evaluator.evaluate_with_intermediate_outputs_continuous_diffusion_analysis(
+                self.cipher, x_inputs, sbox_precomputations, sbox_precomputations_mix_columns)[1]
+
+        if output_dict is None:
+            output_dict = {}
+        for tag_output in output_evaluated_continuous_diffusion_analysis_x.keys():
+            obj = {}
+            x_c_list = output_evaluated_continuous_diffusion_analysis_x[tag_output]
+            obj[tag_output] = [{} for _ in range(len(x_c_list))]
+
+            for j in range(len(x_c_list)):
+                x_c = x_c_list[j]['intermediate_output']
+                output_bits_sum = sum([mag(x_c[idx]) for idx in output_dict[tag_output]])
+                if str(x_c_list[j]['round']) not in obj[tag_output][j]:
+                    obj[tag_output][j][str(x_c_list[j]['round'])] = output_bits_sum
+                else:
+                    obj[tag_output][j][str(x_c_list[j]['round'])] += output_bits_sum
+            sum_by_round.append(obj)
+
+        return sum_by_round
+
+    @staticmethod
+    def _get_graph_representation_components_by_type(graph_representation, type_name):
+        cipher_rounds = graph_representation["cipher_rounds"][0]
+        components_by_type = list(filter(lambda d: d['type'] in [type_name], cipher_rounds))
+
+        return components_by_type
+
+    @staticmethod
+    def _get_graph_representation_tag_output_sizes(graph_representation):
+        temp_components = {}
+        component_descriptions = []
+        cipher_rounds = graph_representation["cipher_rounds"]
+        for cipher_round in cipher_rounds:
+            for component in cipher_round:
+                if (component["type"] == "intermediate_output" or component["type"] == "cipher_output") and \
+                        component["description"] not in component_descriptions:
+                    component_descriptions.append(component["description"])
+
+                    temp_components[component["description"][0]] = list(range(component["output_bit_size"]))
+
+        return temp_components
+
+    def continuous_avalanche_factor(self, lambda_value, number_of_samples):
+        """
+        Continuous generalization of the metric Avalanche Factor. This method implements Definition 14 of [MUR2020]_.
+
+        INPUT:
+
+        - ``lambda_value`` --  **float**; threshold value used to express the input difference
+        - ``number_of_samples`` --  **integer**; number of samples used to compute the continuous avalanche factor
+
+        EXAMPLES::
+
+            sage: from claasp.ciphers.block_ciphers.speck_block_cipher import SpeckBlockCipher as speck
```

Comment:
Needs to be updated to work with the new class

Reply:
done